lib/jsdom/living/nodes/HTMLFormElement-impl.js

```diff
@@ -53,19 +59,102 @@ class HTMLFormElementImpl extends HTMLElementImpl {
     return this.elements.length;
   }
 
-  _doSubmit() {
+  submit() {
+    this._doSubmit(this, {submittedFromSubmit: true});
+  }
+  // https://html.spec.whatwg.org/#concept-form-submit
+  _doSubmit(submitter, flags = {}) {
     if (!isConnected(this)) {
       return;
     }
-    const ev = this._ownerDocument.createEvent("HTMLEvents");
-    ev.initEvent("submit", true, true);
-    if (this.dispatchEvent(ev)) {
-      this.submit();
+
+    const form = this;
+    const document = this._ownerDocument;
+    const window = document._defaultView;
+
+    // NOT IMPLEMENTED: validate the form
+
+    if (!flags.submittedFromSubmit) {
+      const ev = document.createEvent("HTMLEvents");
+      ev.initEvent("submit", true, true);
+      if (this.dispatchEvent(ev)) {
+        return;
+      }
     }
-  }
 
-  submit() {
-    notImplemented("HTMLFormElement.prototype.submit", this._ownerDocument._defaultView);
+    const formDataSet = constructTheFormDataSet(this, submitter);
+
+    const action = this._getAttributeForSubmission('action', submitter) || document.URL;
+    const parsedAction = parseURLToResultingURLRecord(action, document);
+    if (!parsedAction) {
+      return;
+    }
+    const scheme = parsedAction.scheme;
+
+    const method = (this._getAttributeForSubmission('method', submitter) || 'get').toLowerCase();
+
+    const target = this._getAttributeForSubmission('target', submitter) || '_self';
+    const {chosen: targetWindow} = chooseBrowsingContext(window, target);
+    if (!targetWindow) {
+      return;
+    }
+
+    if (method === 'dialog') {
+      notImplemented("dialog method on forms", window);
+    }
+    console.dir({method, action});
+    switch (scheme) {
+      case 'http':
+      case 'https':
+        if (method === 'get') {
+          mutateActionURL();
+        } else {
+          submitAsEntityBody();
+        }
+        break;
+      case 'ftp':
+      case 'javascript':
+        getActionURL();
+        break;
+      case 'data':
+        if (method === 'get') {
+          mutateActionURL();
+        } else {
+          getActionURL();
+        }
+        break;
+      default:
+        notImplemented("url scheme " + scheme, window);
+    }
+    // https://html.spec.whatwg.org/#submit-mutate-action
+    function mutateActionURL() {
```

Comment:
Would be ideal to avoid nested functions where possible, moving these to top-level of the module or class and passing in parameters as appropriate